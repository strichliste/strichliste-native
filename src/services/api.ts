const API_URL = "https://demo.strichliste.org/api/";

export async function fetchJson(
  endpoint: string,
  options: RequestInit = {}
): Promise<any> {
  return fetch(`${API_URL}${endpoint}`, options).then(async res => res.json());
}

export async function post(endpoint: string, params: any): Promise<any> {
  const options: RequestInit = {
    mode: "cors",
    method: "POST",
    body: JSON.stringify(params),
    headers: { "Content-Type": "application/json; charset=utf-8" }
  };
  return fetch(`${API_URL}${endpoint}`, options).then(async res => res.json());
}

export async function get(endpoint: string): Promise<any> {
  return fetch(`${API_URL}${endpoint}`).then(async res => res.json());
}

export async function restDelete(endpoint: string): Promise<any> {
  return fetch(`${API_URL}${endpoint}`, {
    method: "delete"
  }).then(async res => res.json());
}
