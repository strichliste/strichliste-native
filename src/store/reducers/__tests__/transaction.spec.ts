/* eslint-disable @typescript-eslint/no-explicit-any */

import { DeepPartial } from "redux";
import { TransactionTypes, startCreatingTransaction, transaction } from "..";
import { Action } from "../..";
import { get, post, restDelete } from "../../../services/api";
import {
  getTransaction,
  isTransactionDeletable,
  startDeletingTransaction,
  startLoadingTransactions
} from "../transaction";

jest.mock("../../../services/api", () => ({
  get: jest.fn(),
  post: jest.fn(),
  restDelete: jest.fn()
}));

describe("transaction reducer", () => {
  let action: DeepPartial<Action>;

  describe("with non matching action", () => {
    it("returns initial state", () => {
      const initialState = {};
      expect(transaction(undefined, {} as any)).toEqual(initialState);
    });
  });

  describe("with a transactionsLoaded action", () => {
    beforeEach(() => {
      action = {
        type: TransactionTypes.TransactionsLoaded,
        payload: [{ id: 2 }, { id: 3 }]
      };
    });

    it("merges transaction to state", () => {
      const result = transaction(undefined, action as Action);
      expect(result).toMatchSnapshot();
    });
  });
});

describe("action creators", () => {
  describe("startCreatingTransaction", () => {
    beforeEach(() => {
      (post as any).mockImplementationOnce(() =>
        Promise.resolve({ transaction: { id: 5, user: { name: "user" } } })
      );
    });

    it("pushes the transaction and dispatches the new transaction and updated user", async () => {
      const store = getMockStore();
      await startCreatingTransaction(store.dispatch, "2", { amount: 1 });

      expect(post).toHaveBeenCalledWith("user/2/transaction", { amount: 1 });
      expect(store.getActions()).toMatchSnapshot();
    });
  });
  describe("startDeletingTransaction", () => {
    beforeEach(() => {
      (restDelete as any).mockImplementationOnce(() =>
        Promise.resolve({ transaction: { id: 5, user: { name: "user" } } })
      );
    });

    it("updates the transaction and dispatches the new transaction and updated user", async () => {
      const store = getMockStore();
      await startDeletingTransaction(store.dispatch, "2", 4);

      expect(restDelete).toHaveBeenCalledWith("user/2/transaction/4");
      expect(store.getActions()).toMatchSnapshot();
    });
  });
  describe("startLoadingTransactions", () => {
    beforeEach(() => {
      (get as any).mockImplementationOnce(() =>
        Promise.resolve({ transactions: [{ id: 2 }, { id: 3 }] })
      );
    });

    it("pushes the transaction and dispatches the new transaction and updated user", async () => {
      const store = getMockStore();
      await startLoadingTransactions(store.dispatch, "2");

      expect(get).toHaveBeenCalledWith("user/2/transaction?offset=0&limit=5");
      expect(store.getActions()).toMatchSnapshot();
    });

    it("sets the given params", async () => {
      const store = getMockStore();
      await startLoadingTransactions(store.dispatch, "2", 50, 12);

      expect(get).toHaveBeenCalledWith("user/2/transaction?offset=50&limit=12");
    });
  });
});

describe("selectors", () => {
  describe("getTransaction", () => {
    it("returns undefined if no transaction is found", () => {
      expect(
        getTransaction(
          {
            transaction: { 1: { id: 1 } }
          } as any,
          5
        )
      ).toBeUndefined();
    });
    it("returns a matching transaction by id", () => {
      expect(
        getTransaction(
          {
            transaction: { 1: { id: 1 } }
          } as any,
          1
        )
      ).toEqual({ id: 1 });
    });
  });

  describe("isTransactionDeletable", () => {
    it("returns flag if transaction is found", () => {
      expect(
        isTransactionDeletable(
          { transaction: { 1: { id: 1, isDeletable: true } } } as any,
          1
        )
      ).toBeTruthy();
    });
    it("returns false if no transaction is found", () => {
      expect(
        isTransactionDeletable(
          { transaction: { 1: { id: 1, isDeletable: true } } } as any,
          2
        )
      ).toBeFalsy();
    });
  });
});
