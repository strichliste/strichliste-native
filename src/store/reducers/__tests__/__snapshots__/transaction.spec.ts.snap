// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`action creators startCreatingTransaction pushes the transaction and dispatches the new transaction and updated user 1`] = `
Array [
  Object {
    "payload": Object {
      "GLOBAL_LOADER": true,
    },
    "type": "SET_LOADER",
  },
  Object {
    "payload": Object {
      "GLOBAL_ERROR": "",
    },
    "type": "SET_ERROR",
  },
  Object {
    "payload": Object {
      "GLOBAL_LOADER": false,
    },
    "type": "SET_LOADER",
  },
  Object {
    "payload": Object {
      "name": "user",
    },
    "type": "[USER] Loaded_Details",
  },
  Object {
    "payload": Array [
      Object {
        "id": 5,
        "user": Object {
          "name": "user",
        },
      },
    ],
    "type": "TRANSACTIONS_LOADED",
  },
]
`;

exports[`action creators startDeletingTransaction updates the transaction and dispatches the new transaction and updated user 1`] = `
Array [
  Object {
    "payload": Object {
      "GLOBAL_LOADER": true,
    },
    "type": "SET_LOADER",
  },
  Object {
    "payload": Object {
      "GLOBAL_ERROR": "",
    },
    "type": "SET_ERROR",
  },
  Object {
    "payload": Object {
      "GLOBAL_LOADER": false,
    },
    "type": "SET_LOADER",
  },
  Object {
    "payload": Object {
      "name": "user",
    },
    "type": "[USER] Loaded_Details",
  },
  Object {
    "payload": Array [
      Object {
        "id": 5,
        "user": Object {
          "name": "user",
        },
      },
    ],
    "type": "TRANSACTIONS_LOADED",
  },
]
`;

exports[`action creators startLoadingTransactions pushes the transaction and dispatches the new transaction and updated user 1`] = `
Array [
  Object {
    "payload": Object {
      "GLOBAL_LOADER": true,
    },
    "type": "SET_LOADER",
  },
  Object {
    "payload": Object {
      "GLOBAL_ERROR": "",
    },
    "type": "SET_ERROR",
  },
  Object {
    "payload": Object {
      "GLOBAL_LOADER": false,
    },
    "type": "SET_LOADER",
  },
  Object {
    "payload": Array [
      Object {
        "id": 2,
      },
      Object {
        "id": 3,
      },
    ],
    "type": "TRANSACTIONS_LOADED",
  },
]
`;

exports[`transaction reducer with a transactionsLoaded action merges transaction to state 1`] = `
Object {
  "2": Object {
    "id": 2,
  },
  "3": Object {
    "id": 3,
  },
}
`;
